### PR TITLE
[Vault-5736] Add (*Client).WithNamespace() for temporary namespace handling

### PR DESCRIPTION
**Pull request by VinnyHC**
Fixes: #14934 

In the API `Client` should have a method to quickly set (or unset) namespaces on calls. Helpful for calls that are intended to be issued against the [root](https://www.vaultproject.io/docs/enterprise/namespaces#root-only-api-paths).

Current flow for switching namespaces:
``` go
	// Create a Client
	nsClient, err := NewClient(DefaultConfig())
	nsClient.SetNamespace("sample-namespace")

	// Clone the client
	rootClient, err := nsClient.Clone()
	// Clear the namespace
	rootClient.ClearNamespace()
	// Make the call
	rootClient.Sys().Health()
```

Improved flow:
``` go
	// Create a client
	nsClient, err := NewClient(DefaultConfig())
	nsClient.SetNamespace("my-ns")
	
	// issued without a namespace
	nsClient.WithNamespace("").Sys().Health()
	
	// issued with X-Vault-Namespace: other-ns
	nsClient.WithNamespace("other-ns").Logical.Read()
	
	// nsClient retains original namespace
	nsClient.Logical.Read()
```

### CODE REVIEW

File: api/client.go [hunk 1]
```diff
@@ -1143,12 +1172,18 @@ func (c *Client) rawRequestWithContext(ctx context.Context, r *Request) (*Respon
...
+	headers := c.headers
 	outputCurlString := c.config.OutputCurlString
 	logger := c.config.Logger
 	c.config.modifyLock.RUnlock()
 
 	c.modifyLock.RUnlock()
 
+	// explicitly set the namespace header to the current client
+	if headers != nil {
+		r.Headers.Set(consts.NamespaceHeaderName, headers.Get(consts.NamespaceHeaderName))
```
Comment (averche) [inline api/client.go:1184]: Should we check that `NamespaceHeaderName` key exists. I'm not sure if inserting an empy string header is equivalent to no header.

File: api/client.go [hunk 2]
```diff
@@ -820,10 +820,39 @@ func (c *Client) setNamespace(namespace string) {
...
+
+// WithNamespace makes a shallow copy of Client, modifies it to use
+// the given namespace, and returns it. Passing an empty string will
+// temporarily unset the namespace.
+func (c *Client) WithNamespace(namespace string) *Client {
+	c2 := *c
+	c2.modifyLock = sync.RWMutex{}
+	c2.headers = c.Headers()
+	if namespace == "" {
+		c2.ClearNamespace()
```
[the remark this answers is not in the document]
Comment (VinnyHC) [reply api/client.go:851]: Yeah, I kicked around not allowing an empty string and have another method like `(*client).WithRootNamespace()` but in the end, I believe our docs and such indicate that setting any namespace variable (env var or client.SetNamespace) with an empty string would indicate the root namespace should be used. And as you mentioned, we don't have a great way to fail out of this call (e.g. returning an error).

File: api/client.go [hunk 3]
```diff
@@ -820,10 +820,39 @@ func (c *Client) setNamespace(namespace string) {
...
+		return ""
+	}
+	return c.headers.Get(consts.NamespaceHeaderName)
+}
+
+// WithNamespace makes a shallow copy of Client, modifies it to use
+// the given namespace, and returns it. Passing an empty string will
+// temporarily unset the namespace.
+func (c *Client) WithNamespace(namespace string) *Client {
+	c2 := *c
```
[the remark this answers is not in the document]
Comment (VinnyHC) [reply api/client.go:847]: This approach was inspired by the `WithRequestCallbacks` and `WithResponseCallbacks` which copy the client in the same manner and replace the lock.

File: api/client.go (hunk 1 above)
Comment (VinnyHC) [reply api/client.go:1184]: Yeah I think that's good move.

File: api/client.go [hunk 4]
```diff
@@ -820,10 +820,39 @@ func (c *Client) setNamespace(namespace string) {
...
+	c2.modifyLock = sync.RWMutex{}
+	c2.headers = c.Headers()
```
[the remark this answers is not in the document]
Comment (VinnyHC) [reply api/client.go:849]: I believe it is. `c.Headers` returns a copy and we're effectively manipulating the headers with these namespace calls, so we want to make sure we're manipulating a new map.

File: api/client.go [hunk 5]
```diff
@@ -1143,12 +1172,20 @@ func (c *Client) rawRequestWithContext(ctx context.Context, r *Request) (*Respon
...
 	backoff := c.config.Backoff
 	httpClient := c.config.HttpClient
+	ns := c.headers.Get(consts.NamespaceHeaderName)
 	outputCurlString := c.config.OutputCurlString
 	logger := c.config.Logger
 	c.config.modifyLock.RUnlock()
 
 	c.modifyLock.RUnlock()
 
+	switch ns {
```
Comment (ncabatoff) [inline api/client.go:1182]: I don't understand why this change should be needed, can you explain please?

Comment (VinnyHC) [reply api/client.go:1182]: Absolutely. When I was writing the tests I came across this pattern:
```go
	_, err = client.WithNamespace("ns1").rawRequestWithContext(
		context.Background(),
		client.NewRequest(http.MethodGet, "/"))
```
When we call `client.NewRequest()` the request will get the headers set to whatever is in `client` ([this code](https://github.com/hashicorp/vault/blob/main/api/client.go#L1101) I believe). This change will make sure at the time of the call the request has the appropriate header map.

File: api/client.go [hunk 6]
```diff
@@ -1280,13 +1317,19 @@ func (c *Client) httpRequestWithContext(ctx context.Context, r *Request) (*Respo
...
 	httpClient := c.config.HttpClient
 	outputCurlString := c.config.OutputCurlString
+	// add headers
 	if c.headers != nil {
 		for header, vals := range c.headers {
 			for _, val := range vals {
 				req.Header.Add(header, val)
 			}
 		}
+		// explicitly set the namespace header to current client
```
[the remark this answers is not in the document]
Comment (VinnyHC) [reply api/client.go:1327]: Similar to your other question, since this uses `(*Request).Header.Add()` which will append all of the headers to the map I wanted to make sure that only one namespace header is set.

**Review by ncabatoff — APPROVED** [review]
